flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java

```diff
@@ -0,0 +1,563 @@
+                return join.copy(join.getTraitSet(), Arrays.asList(newProbe, buildSide));
+            }
+        }
+
+        return join;
+    }
+
+    /**
+     * Inject runtime filter and return the new probe side (without exchange).
+     *
+     * @param buildSide the build side
+     * @param probeSide the probe side
+     * @param buildIndices the build projection
+     * @param probeIndices the probe projection
+     * @return the new probe side
+     */
+    private static RelNode createNewProbeWithRuntimeFilter(
+            RelNode buildSide,
+            RelNode probeSide,
+            ImmutableIntList buildIndices,
+            ImmutableIntList probeIndices) {
+        Optional<Double> buildRowCountOpt = getEstimatedRowCount(buildSide);
+        checkState(buildRowCountOpt.isPresent());
+        int buildRowCount = buildRowCountOpt.get().intValue();
+        int maxRowCount =
+                (int)
+                        Math.ceil(
+                                getMaxBuildDataSize(buildSide)
+                                        / FlinkRelMdUtil.binaryRowAverageSize(buildSide));
+        double filterRatio = computeFilterRatio(buildSide, probeSide, buildIndices, probeIndices);
+
+        RelNode localBuilder =
+                new BatchPhysicalLocalRuntimeFilterBuilder(
+                        buildSide.getCluster(),
+                        buildSide.getTraitSet(),
+                        buildSide,
+                        buildIndices.toIntArray(),
+                        buildRowCount,
+                        maxRowCount);
+        RelNode globalBuilder =
+                new BatchPhysicalGlobalRuntimeFilterBuilder(
+                        localBuilder.getCluster(),
+                        localBuilder.getTraitSet(),
+                        createExchange(localBuilder, FlinkRelDistribution.SINGLETON()),
+                        maxRowCount);
+        RelNode runtimeFilter =
+                new BatchPhysicalRuntimeFilter(
+                        probeSide.getCluster(),
+                        probeSide.getTraitSet(),
+                        createExchange(globalBuilder, FlinkRelDistribution.BROADCAST_DISTRIBUTED()),
+                        probeSide,
+                        probeIndices.toIntArray(),
+                        filterRatio);
+
+        return runtimeFilter;
+    }
+
+    /**
+     * Find a suitable build side. In order not to affect MultiInput, when the original build side
+     * of runtime filter is not an {@link Exchange}, we need to push down the builder, until we find
+     * an exchange and inject the builder there.
+     *
+     * @param rel the original build side
+     * @param buildIndices build indices
+     * @param buildSideChecker check whether current build side is suitable
+     * @return An optional info of the suitable build side.It will be empty if we cannot find the
+     *     suitable build side.
+     */
+    private static Optional<BuildSideInfo> findSuitableBuildSide(
```

Comment:
Can you give more explanation of this method's logic? Especially the various if else branch.